### PR TITLE
pdn: account for vias in macro pins

### PR DESCRIPTION
**Pull request by gadfort**
This should address #2650

### CODE REVIEW

**Comment by gadfort** [comment]
@oharboe FYI, this should solve the issue in PDN, if you have time, please check

**Comment by oharboe** [comment]
> @oharboe FYI, this should solve the issue in PDN, if you have time, please check

What about all the other cases I discovered in my pr, including the openroad gui?

**Comment by gadfort** [comment]
> > @oharboe FYI, this should solve the issue in PDN, if you have time, please check
> 
> What about all the other cases I discovered in my pr, including the openroad gui?

@oharboe I think you should consider opening separate issues for each one you encountered. Feel free to assign me to the GUI related one.

File: src/pdn/src/grid.cpp [hunk 1]
```diff
@@ -1260,12 +1260,28 @@ ShapeTreeMap InstanceGrid::getInstancePins(odb::dbInst* inst)
...
-          transform.apply(box_rect);
-          auto shape
-              = std::make_shared<Shape>(box->getTechLayer(), net, box_rect);
-          shape->setShapeType(Shape::FIXED);
-          pins.push_back(shape);
+          if (box->isVia()) {
+            odb::dbTechVia* tech_via = box->getTechVia();
+            if (tech_via == nullptr) {
+              continue;
+            }
```
Comment (maliberty) [inline src/pdn/src/grid.cpp:1267]: What is the have a VIA from the LEF itself and not the tech?

Comment (gadfort) [reply src/pdn/src/grid.cpp:1267]: Would that manifest as a block via? I think there only those two, correct?

**Comment by QuantamHD** [comment]
@maliberty @gadfort Should we consider modifying the ODB API to force users to deal with this? It seems like the current API is ripe with "shooting yourself in the foot".

**Comment by maliberty** [comment]
> @maliberty @gadfort Should we consider modifying the ODB API to force users to deal with this? It seems like the current API is ripe with "shooting yourself in the foot".

What change do you suggest?

**Comment by oharboe** [comment]
@gadfort I've created a pull request that you can use to reproduce the problem: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/709

**Comment by QuantamHD** [comment]
> > @maliberty @gadfort Should we consider modifying the ODB API to force users to deal with this? It seems like the current API is ripe with "shooting yourself in the foot".
> 
> What change do you suggest?

Perhaps `auto* box : mpin->getGeometry(); box->getBox()` should be removed in favor of a more general getBoxes function that returned a list of shapes in that box. Which could automatically do the kind of thing Peter did in this PR. 

The other thing that might work is having a function that returns std::variant<dbTechVia*, db::Rect*, ...> that would force users to check on the type.

File: src/pdn/src/grid.cpp (hunk 1 above)
Comment (maliberty) [reply src/pdn/src/grid.cpp:1267]: I see that in lefin any via would become a tech via.  So I guess it isn't possible to have a block via in a macro pin.

Yes there are only block & tech vias.

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
Based on the comments in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/709 it seems this does resolve this issue reported by @oharboe

**Pull request merged**